Added new test test_compare_cpu that checks if cpu and gpu results are consistent

srossross opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #85583
* #86462
* __->__ #85011

srossross committed:
Added new test test_compare_cpu that checks if cpu and gpu results are consistent

[ghstack-poisoned]

srossross committed:
Update on "Added new test test_compare_cpu that checks if cpu and gpu results are consistent"




[ghstack-poisoned]

mruberry commented on test/test_ops.py:
```diff
@@ -160,6 +160,43 @@ def test_numpy_ref(self, device, dtype, op):
+    @ops(_ops_and_refs)
+    def test_compare_cpu(self, device, dtype, op):
+
+        if dtype not in op.dtypes:
```
Running this test in every supported dtype seems like a lot. Maybe just once is OK to start with?  The test can ask for "any one" dtype following the instructions here:

https://github.com/pytorch/pytorch/blob/ebd4e90ff7f6d1e57bbf6f2e717a8addb6b28e28/torch/testing/_internal/common_device_type.py#L674

srossross replied:
done

mruberry commented on test/test_ops.py:
```diff
@@ -160,6 +160,43 @@ def test_numpy_ref(self, device, dtype, op):
+            cuda_results = op(sample.input, *sample.args, **sample.kwargs)
+            cpu_results = op(cpu_sample.input, *cpu_sample.args, **cpu_sample.kwargs)
+
+            self.assertEqual(type(cuda_results), type(cpu_results))
```
This may want to call `assert_close` directly:

https://pytorch.org/docs/master/testing.html#torch.testing.assert_close

Which will verify the same type is returned

mruberry commented on test/test_ops.py:
```diff
@@ -160,6 +160,43 @@ def test_numpy_ref(self, device, dtype, op):
+            # Simple first check that the return values have the same number of arguments
+            self.assertEqual(len(cuda_results), len(cpu_results))
+
+            for cpu_result, cuda_result in zip(cpu_results, cuda_results):
```
Just calling `assert_close` with `check_device=False` is probably all that needs to be done?

mruberry commented on test/test_ops.py:
```diff
@@ -160,6 +160,43 @@ def test_numpy_ref(self, device, dtype, op):
+    # Tests that the cpu and gpu results are consistent
+    @onlyCUDA
+    @suppress_warnings
+    @ops(_ops_and_refs)
```
This test should only run for operators that do not have Python references, since there's already a tests that checks for consistency with those

srossross replied:
I removed all ops where `op.ref is not None`

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -14802,6 +14802,8 @@ def reference_flatten(input, start_dim=0, end_dim=-1):
             DecorateInfo(unittest.expectedFailure, 'TestJit', 'test_variant_consistency_jit'),
             # Skip Nvfuser
             DecorateInfo(unittest.skip('Skipped!'), 'TestCudaFuserOpInfo'),
+            # RuntimeError: t == DeviceType::CUDA INTERNAL ASSERT FAILED
```
Have the Jiterated ops fail this new test seems conceptually odd -- they should probably be skipping it instead, and without requiring an explicit skip. Maybe their `dtypes` needs to change to `dtypesIfCUDA` to signify they don't support CPU inputs better?

srossross replied:
This is decorated with :

```
decorators=[
            onlyCUDA,
```

Is there a way I can use this?
Another possibility is change the comment to just `# This is not implemented on CPU`  if `t == DeviceType::CUDA ` is not important.

mruberry replied:
We could think about reusing the decorator, but I think the most natural way for an op to signal it only supports the CUDA device type is to define `dtypes` as empty and then define `dtypesIfCUDA`

srossross replied:
This test ended up being skipped because the numpy `ref=` is defined

srossross committed:
Update on "Added new test test_compare_cpu that checks if cpu and gpu results are consistent"




[ghstack-poisoned]

srossross committed:
Update on "Added new test test_compare_cpu that checks if cpu and gpu results are consistent"




[ghstack-poisoned]

srossross committed:
Update on "Added new test test_compare_cpu that checks if cpu and gpu results are consistent"




[ghstack-poisoned]

srossross commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -9342,6 +9347,8 @@ def reference_flatten(input, start_dim=0, end_dim=-1):
                DecorateInfo(unittest.skip("Skipped!"), 'TestGradients', 'test_fn_gradgrad'),
                # GradcheckError: gradcheck expects all tensor inputs are dense when check_sparse_nnz is set to False
                DecorateInfo(unittest.skip("Skipped!"), 'TestGradients', 'test_forward_mode_AD'),
+               # Some inputs are non deterministic, eg when mat1 and mat2 are size 0
+               DecorateInfo(unittest.skip('some inputs are non deterministic'), 'TestCommon', 'test_compare_cpu'),
```
@amjames here is the case that is failing: `1,1,1,1,1 != 0.2,0.2,0.2,0.2,0.2`  

```py
In [19]: torch.sparse.sampled_addmm(torch.eye(5).to_sparse_csr(), torch.zeros([5,0]), torch.zeros([0,5]), alpha=0.6, beta=0.2)
Out[19]: 
tensor(crow_indices=tensor([0, 1, 2, 3, 4, 5]),
       col_indices=tensor([0, 1, 2, 3, 4]),
       values=tensor([0.2000, 0.2000, 0.2000, 0.2000, 0.2000]), size=(5, 5),
       nnz=5, layout=torch.sparse_csr)

In [20]: torch.sparse.sampled_addmm(torch.eye(5, device='cuda:0').to_sparse_csr(), torch.zeros([5,0], device='cuda:0'), torch.zeros([0,5], device='cuda:0'), alpha=0.6, beta=0.2)
Out[20]: 
tensor(crow_indices=tensor([0, 1, 2, 3, 4, 5]),
       col_indices=tensor([0, 1, 2, 3, 4]),
       values=tensor([1., 1., 1., 1., 1.]), device='cuda:0', size=(5, 5),
       nnz=5, layout=torch.sparse_csr)
```

amjames replied:
This is a real bug in the cuda implementation. You should open an issue.

srossross committed:
Update on "Added new test test_compare_cpu that checks if cpu and gpu results are consistent"




[ghstack-poisoned]

srossross committed:
Update on "Added new test test_compare_cpu that checks if cpu and gpu results are consistent"




[ghstack-poisoned]

srossross committed:
Update on "Added new test test_compare_cpu that checks if cpu and gpu results are consistent"




[ghstack-poisoned]

srossross commented on torch/testing/_internal/opinfo/definitions/fft.py:
```diff
@@ -217,7 +221,18 @@ def mt(shape, **kwargs):
                 "test_reference_nd",
-            )
+            ),
+            DecorateInfo(
+                toleranceOverride(
+                    {
+                        torch.float16: tol(atol=1e-4, rtol=1e-4),
+                        torch.complex64: tol(atol=1e-4, rtol=1e-4),
```
@amjames any idea why this test is failing? 

https://github.com/pytorch/pytorch/actions/runs/3070095254/jobs/4959762987

```
======================================================================
FAIL [0.007s]: test_compare_cpu_fft_hfft2_cuda_float32 (__main__.TestCommonCUDA)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/conda/lib/python3.10/site-packages/torch/testing/_internal/common_utils.py", line 1950, in wrapper
    method(*args, **kwargs)
  File "/opt/conda/lib/python3.10/site-packages/torch/testing/_internal/common_device_type.py", line 378, in instantiated_test
    result = test(self, **param_kwargs)
  File "/opt/conda/lib/python3.10/site-packages/torch/testing/_internal/common_device_type.py", line 853, in dep_fn
    return fn(slf, *args, **kwargs)
  File "/opt/conda/lib/python3.10/site-packages/torch/testing/_internal/common_device_type.py", line 815, in test_wrapper
    return test(*args, **kwargs)
  File "/opt/conda/lib/python3.10/site-packages/torch/testing/_internal/common_device_type.py", line 975, in only_fn
    return fn(slf, *args, **kwargs)
  File "/opt/conda/lib/python3.10/site-packages/torch/testing/_internal/common_utils.py", line 1283, in wrapper
    fn(*args, **kwargs)
  File "/var/lib/jenkins/workspace/test/test_ops.py", line 197, in test_compare_cpu
    self.assertEqual(cpu_result, cuda_result)
  File "/opt/conda/lib/python3.10/site-packages/torch/testing/_internal/common_utils.py", line 2411, in assertEqual
    assert_equal(
  File "/opt/conda/lib/python3.10/site-packages/torch/testing/_comparison.py", line 1093, in assert_equal
    raise error_metas[0].to_error(msg)
AssertionError: Tensor-likes are not close!

Mismatched elements: 3 / 360 (0.8%)
Greatest absolute difference: 2.288818359375e-05 at index (2, 5, 10) (up to 1e-05 allowed)
Greatest relative difference: 5.305602716468591e-05 at index (1, 0, 3) (up to 1.3e-06 allowed)
```
the atol and rtol look to be within the limits I have set here

And it passes on my machine

srossross committed:
Update on "Added new test test_compare_cpu that checks if cpu and gpu results are consistent"




[ghstack-poisoned]

srossross committed:
Update on "Added new test test_compare_cpu that checks if cpu and gpu results are consistent"




[ghstack-poisoned]

srossross committed:
Update on "Added new test test_compare_cpu that checks if cpu and gpu results are consistent"




[ghstack-poisoned]

srossross committed:
Update on "Added new test test_compare_cpu that checks if cpu and gpu results are consistent"




[ghstack-poisoned]

srossross commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -15317,6 +15317,8 @@ def reference_flatten(input, start_dim=0, end_dim=-1):
             # lambda impl
             DecorateInfo(unittest.expectedFailure, 'TestJit', 'test_variant_consistency_jit'),
             DecorateInfo(unittest.expectedFailure, 'TestNormalizeOperators', 'test_normalize_operator_exhaustive'),
+            # FIXME: this fails on CI and passes on quansight's qgpu machine
+            DecorateInfo(unittest.expectedFailure, 'TestCommon', 'test_compare_cpu'),
```
@amjames The tests are passing after this `expectedFailure` which fails on qgpu with 

```
================================================== 434 passed, 501 skipped, 47841 deselected, 26 xfailed, 11 warnings in 34.78s ==================================================
(pytorch2) srossross@qgpu1:~/pytorch$ py.test test/test_ops.py -v -k test_compare_cpu_nn_functional_embedding_cuda_float32
============================================================================== test session starts ===============================================================================
platform linux -- Python 3.8.13, pytest-7.1.2, pluggy-1.0.0 -- /home/srossross/.conda/envs/pytorch2/bin/python
cachedir: .pytest_cache
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase('/home/srossross/pytorch/.hypothesis/examples')
rootdir: /home/srossross/pytorch, configfile: pytest.ini
plugins: hypothesis-6.54.4
collected 48802 items / 48801 deselected / 1 selected                                                                                                                            

test/test_ops.py::TestCommonCUDA::test_compare_cpu_nn_functional_embedding_cuda_float32 FAILED                                                                             [100%]

==================================================================================== FAILURES ====================================================================================
______________________________________________________ TestCommonCUDA.test_compare_cpu_nn_functional_embedding_cuda_float32 ______________________________________________________
Unexpected success
============================================================================ short test summary info =============================================================================
FAILED test/test_ops.py::TestCommonCUDA::test_compare_cpu_nn_functional_embedding_cuda_float32
=============================================================== 1 failed, 48801 deselected, 11 warnings in 15.09s ================================================================
```

amjames replied:
cc @lezcano

srossross replied:
[here is](https://github.com/pytorch/pytorch/actions/runs/3071172014/jobs/4986626966) the original error 
```
======================================================================
FAIL [0.009s]: test_compare_cpu_nn_functional_embedding_cuda_float32 (__main__.TestCommonCUDA)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/conda/lib/python3.10/site-packages/torch/testing/_internal/common_utils.py", line 1950, in wrapper
    method(*args, **kwargs)
  File "/opt/conda/lib/python3.10/site-packages/torch/testing/_internal/common_device_type.py", line 378, in instantiated_test
    result = test(self, **param_kwargs)
  File "/opt/conda/lib/python3.10/site-packages/torch/testing/_internal/common_device_type.py", line 815, in test_wrapper
    return test(*args, **kwargs)
  File "/opt/conda/lib/python3.10/site-packages/torch/testing/_internal/common_device_type.py", line 975, in only_fn
    return fn(slf, *args, **kwargs)
  File "/opt/conda/lib/python3.10/site-packages/torch/testing/_internal/common_utils.py", line 1283, in wrapper
    fn(*args, **kwargs)
  File "/var/lib/jenkins/workspace/test/test_ops.py", line 187, in test_compare_cpu
    self.assertEqual(cuda_results, cpu_results)
  File "/opt/conda/lib/python3.10/site-packages/torch/testing/_internal/common_utils.py", line 2412, in assertEqual
    assert_equal(
  File "/opt/conda/lib/python3.10/site-packages/torch/testing/_comparison.py", line 1093, in assert_equal
    raise error_metas[0].to_error(msg)
AssertionError: Tensor-likes are not close!

Mismatched elements: 15 / 45 (33.3%)
Greatest absolute difference: 0.5051435921341181 at index (0, 0, 3) (up to 1e-05 allowed)
Greatest relative difference: 0.9674892692698572 at index (0, 0, 2) (up to 1.3e-06 allowed)
```

lezcano replied:
I'd skip it if it's not clear where it fails and I'd open an issue and point the skip to the issue.

Could it be the case that this function is not deterministic for some inputs?

srossross replied:
added [issue ](https://github.com/pytorch/pytorch/issues/85377)

srossross commented on test/test_ops.py:
```diff
@@ -160,6 +160,43 @@ def test_numpy_ref(self, device, dtype, op):
+                cpu_results = (cpu_results,)
+
+            # Simple first check that the return values have the same number of arguments
+            self.assertEqual(len(cuda_results), len(cpu_results))
```
simplified

srossross committed:
Update on "Added new test test_compare_cpu that checks if cpu and gpu results are consistent"




[ghstack-poisoned]

srossross committed:
Update on "Added new test test_compare_cpu that checks if cpu and gpu results are consistent"




[ghstack-poisoned]

srossross committed:
Update on "Added new test test_compare_cpu that checks if cpu and gpu results are consistent"




[ghstack-poisoned]

amjames commented on torch/testing/_internal/opinfo/definitions/linalg.py:
```diff
@@ -1450,6 +1450,12 @@ def make_input():
                 device_type="mps",
                 dtypes=[torch.float32],
             ),
+            # This appears to be a different algorithm on cpu vs gpu
```
I am curious why `eig` is a skip but later on you mark `eigh` xfail?

srossross replied:
updated the comment, This will fail on CI, but locally passes with a newer cuda toolkit version

amjames commented on torch/testing/_internal/opinfo/definitions/linalg.py:
```diff
@@ -1657,6 +1665,11 @@ def make_input():
                 dtypes=[torch.float32],
             ),
+            DecorateInfo(
+                unittest.skip("cpu vs gpu is sometimes the same"),
```
NIT: Although the meaning is the same, should probably read `sometimes different`

srossross replied:
updated

amjames commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -8190,6 +8190,8 @@ def reference_flatten(input, start_dim=0, end_dim=-1):
                # aten.uniform was not decomposed
                DecorateInfo(unittest.expectedFailure, 'TestDecomp', 'test_quick'),
+               # Can not test non deterministic ops between devices
+               DecorateInfo(unittest.expectedFailure, 'TestCommon', 'test_compare_cpu'),
```
The expected failures due to non-deterministic behavior should probably all be `skip` as coincidental successes would cause the test to be flaky. You have done this in some places but many still use xfail.

amjames reviewed (changes requested):
Looks very good! Just have one nit, and there are some inconsistencies between the application of skips vs xfails. 


See inline comments.

srossross committed:
Update on "Added new test test_compare_cpu that checks if cpu and gpu results are consistent"




[ghstack-poisoned]

lezcano commented on torch/testing/_internal/opinfo/definitions/linalg.py:
```diff
@@ -1657,6 +1665,11 @@ def make_input():
                 device_type="mps",
                 dtypes=[torch.float32],
             ),
+            DecorateInfo(
+                unittest.skip("cpu vs gpu is sometimes different"),
+                "TestCommon",
+                "test_compare_cpu",
+            ),
```
Yeah, `lstsq` is a tricky one. In particular, it's known to be non-deterministic on CPU. See https://github.com/pytorch/pytorch/issues/71222. Can you mention this issue in the skip?

lezcano commented on torch/testing/_internal/opinfo/definitions/linalg.py:
```diff
@@ -2259,6 +2284,8 @@ def make_input():
                 device_type="mps",
                 dtypes=[torch.float32],
             ),
+            # This appears to be a different algorithm on cpu vs gpu
+            DecorateInfo(unittest.expectedFailure, "TestCommon", "test_compare_cpu"),
```
This is a similar issue to that of `eigh`. If you postprocess the output of the function with `output_process_fn_grad` in the test, the error should go.

lezcano reviewed (commented):
A few notes on the linalg issues.

srossross commented on torch/testing/_internal/opinfo/definitions/linalg.py:
```diff
@@ -1529,6 +1535,8 @@ def make_input():
                 device_type="mps",
                 dtypes=[torch.float32],
             ),
+            # linalg.eigh appears to be a different algorithm on cpu vs gpu
+            DecorateInfo(unittest.expectedFailure, "TestCommon", "test_compare_cpu"),
```
@lezcano I'm not sure I understand what you are proposing. 

The results show that col 3 is off by a factor of `-1`

```py
(Pdb) cuda_results.eigenvectors
tensor([[-0.0258, -0.4235,  0.7116, -0.3142,  0.4636],
        [-0.4843, -0.6299, -0.2308, -0.3176, -0.4633],
        [-0.4982, -0.0339,  0.3929,  0.7578, -0.1482],
        [ 0.0720,  0.4078,  0.5271, -0.3370, -0.6611],
        [-0.7151,  0.5065, -0.0901, -0.3355,  0.3337]], device='cuda:0')
(Pdb) cpu_results.eigenvectors
tensor([[-0.0258, -0.4235,  0.7116,  0.3142,  0.4636],
        [-0.4843, -0.6299, -0.2308,  0.3176, -0.4633],
        [-0.4982, -0.0339,  0.3929, -0.7578, -0.1482],
        [ 0.0720,  0.4078,  0.5271,  0.3370, -0.6611],
        [-0.7151,  0.5065, -0.0901,  0.3355,  0.3337]])
```

Looking at other functions using `output_process_fn_grad` I tried `sample.output_process_fn_grad(results)` which has the same issue:

```
(Pdb) cpu_sample.output_process_fn_grad(cpu_results.eigenvectors)
tensor([[-0.0258, -0.4235,  0.7116,  0.3142,  0.4636],
        [-0.4843, -0.6299, -0.2308,  0.3176, -0.4633],
        [-0.4982, -0.0339,  0.3929, -0.7578, -0.1482],
        [ 0.0720,  0.4078,  0.5271,  0.3370, -0.6611],
        [-0.7151,  0.5065, -0.0901,  0.3355,  0.3337]])
(Pdb) sample.output_process_fn_grad(cuda_results.eigenvectors)
tensor([[-0.0258, -0.4235,  0.7116, -0.3142,  0.4636],
        [-0.4843, -0.6299, -0.2308, -0.3176, -0.4633],
        [-0.4982, -0.0339,  0.3929,  0.7578, -0.1482],
        [ 0.0720,  0.4078,  0.5271, -0.3370, -0.6611],
        [-0.7151,  0.5065, -0.0901, -0.3355,  0.3337]], device='cuda:0')
```

lezcano replied:
`output_process_fn_grad` needs to process the output of the whole function. See the definition of this function in the function that samples the values for the OpInfo.
The eigenvectors are defined up to multiplication by -1 in the real case, and up to multiplication by e^{i\phi} (a number of norm 1) in the complex case, hence this discrepancy, and that's why that `output_process_fn_grad` solves it.

srossross replied:
Should I try to apply `output_process_fn_grad ` to all tests?

lezcano replied:
I believe so, yes.

srossross committed:
Update on "Added new test test_compare_cpu that checks if cpu and gpu results are consistent"




[ghstack-poisoned]

srossross commented on test/test_ops.py:
```diff
@@ -160,6 +161,34 @@ def test_numpy_ref(self, device, dtype, op):
+
+        for sample in samples:
+            cpu_sample = sample.transform(to_cpu)
+            cuda_results = op(sample.input, *sample.args, **sample.kwargs)
+            cpu_results = op(cpu_sample.input, *cpu_sample.args, **cpu_sample.kwargs)
+
+            cuda_results = sample.output_process_fn_grad(cuda_results)
+            cpu_results = cpu_sample.output_process_fn_grad(cpu_results)
```
@lezcano is this what you had in mind?

lezcano replied:
Yes! Thank you!

lezcano replied:
Happy to see that we could get rid of all those xfails :)

lezcano commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -12240,7 +12270,12 @@ def reference_flatten(input, start_dim=0, end_dim=-1):
+           decorators=[
+               skipCUDAIfNoCusolver,
+               skipCPUIfNoLapack,
+               # Resulting values are very large therefore atol is increased and rely on rtol
+               DecorateInfo(toleranceOverride({torch.float32: tol(atol=5000, rtol=1e-05)}),
+                            'TestCommon', 'test_compare_cpu')],
```
This doesn't look right. If you increase the atol so much it's the same as testing nothing.

srossross replied:
Yes, This was to turn the atol 'off', the elements in question are roughly `1**10` so I was relying on the rtol.

* cpu   `3.71387269e+10`
* cuda `3.71387310e+10`

how should I proceed?

lezcano replied:
then you should do `atol=0`, right?

srossross replied:
I thought atol and rtol were mutually exclusive. fixed it now

lezcano reviewed (commented):
Just one small point. At any rate, it's good to see that there's not many functions that have intrinsic issues :)

lezcano commented:
Just one point. There are a few functions that are noted as non-deterministic. Now, the fact that the function is non-deterministic for some inputs doesn't mean that it's non-deterministic for all inputs. Some of course are (like `normal`) but others like `index_put_(accumulate=False)` is only non-deterministic when there are repeated indices. Same happens with a unstable sort and an input with repeated elements.
Usually, in OpInfos, we take care of not generating these non-deterministic inputs, because these functions are not differentiable on these inputs, and gradtests would fail on them. As such my question is, are all the functions that are marked as skips coming from non-determinism non-deterministic for all inputs, or is there any of them that's just non-deterministic for some inputs. If it's the latter, it may be the case that the function has some deeper issue as the gradtests would be failing for them.

srossross committed:
Update on "Added new test test_compare_cpu that checks if cpu and gpu results are consistent"




[ghstack-poisoned]

srossross commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -10451,7 +10459,10 @@ def reference_flatten(input, start_dim=0, end_dim=-1):
                DecorateInfo(unittest.skip('Passes on complex128 and float64 only'), 'TestGradients', 'test_fn_fwgrad_bwgrad'),
                # AssertionError: Tensor-likes are not close! (new_empty_strided.default)
-               DecorateInfo(unittest.skip("Expected: new_empty_strided is not comparable"), 'TestDecomp', 'test_comprehensive'),)),
+               DecorateInfo(unittest.skip("Expected: new_empty_strided is not comparable"), 'TestDecomp', 'test_comprehensive'),
+               DecorateInfo(
+                   unittest.skip("Some stride values write multiple values to the same location e.g. (1,1,1,1)"),
+                   'TestCommon', 'test_compare_cpu'),)),
```
@lezcano this is the only op with inputs that can be non-deterministic

lezcano replied:
Great, thanks! A good follow-up PR would be to see if you can generate inputs that do not yield non-deterministic behaviour. I wouldn't be surprised if that fixed a few of the other skip/errors this function has.

srossross replied:
I created a WIP here https://github.com/pytorch/pytorch/pull/85583

srossross committed:
Update on "Added new test test_compare_cpu that checks if cpu and gpu results are consistent"




[ghstack-poisoned]

amjames commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -9505,6 +9506,8 @@ def reference_flatten(input, start_dim=0, end_dim=-1):
                # GradcheckError: gradcheck expects all tensor inputs are dense when check_sparse_nnz is set to False
                DecorateInfo(unittest.skip("Skipped!"), 'TestGradients', 'test_forward_mode_AD'),
+               # https://github.com/pytorch/pytorch/issues/85169
+               DecorateInfo(unittest.expectedFailure, 'TestCommon', 'test_compare_cpu'),
```
You should be able to remove this xfail now that the stack has been re-ordered. `sparse.sampled_addmm` should not still be failing, if it is there is another bug.

amjames replied:
Yes, it should be removed the unexpected success is causing the failure of ([linux-bionic-cuda11.6-py3.10-gcc7 / test (default, 1, 4, linux.4xlarge.nvidia.gpu)](https://github.com/pytorch/pytorch/actions/runs/3108192266/jobs/5037447741))

srossross replied:
removed

srossross committed:
Update on "Added new test test_compare_cpu that checks if cpu and gpu results are consistent"




[ghstack-poisoned]

lezcano commented on test/test_ops.py:
```diff
@@ -160,6 +161,34 @@ def test_numpy_ref(self, device, dtype, op):
+    # Tests that the cpu and gpu results are consistent
+    @onlyCUDA
+    @suppress_warnings
+    @ops(_ops_and_refs_with_no_numpy_ref, dtypes=OpDTypes.any_one)
+    def test_compare_cpu(self, device, dtype, op):
+
+        if dtype not in op.dtypes:
+            raise unittest.SkipTest("This test requires both cuda and cpu support")
+        if dtype not in op.dtypesIfCUDA:
+            raise unittest.SkipTest("This test requires both cuda and cpu support")
```
`any_one` will give the first dtype available for, in this case, CUDA. That does not mean that there isn't a dtype in the intersection of CPU and CUDA dtypes.

amjames replied:
A potential workaround: 

```python
    @ops(_ops_and_refs_with_no_numpy_ref, dtypes=OpDTypes.none)
    def test_compare_cpu(self, device, dtype, op):
        dtype_choices = op.dtypes.intersection(op.dtypesIfCUDA)
        if len(dtype_choices) == 0: 
            raise unittest.SkipTest("This test requires both cuda and cpu support")
        dtype = next(iter(dtype_choices)) # or dtype_choices.pop() as you like
```

Basically you would ensure the test is only instantiated for single `dtype` value (namely `dtype=None`) and then find an appropriate `dtype` for the test if one exists internally.

mruberry replied:
Isn't that what `OpDtypes.any_one` does?

I think this check is redundant with the requested `dtypes`.

srossross replied:
After talking with amjames, I updated the comment. any_one chooses `float32` if available, please let me know if you want me to update with the proposed solution

srossross replied:
After trying this out, it seems like `toleranceOverride` does not work with `dtypes=OpDTypes.none`

lezcano replied:
Yeah, that's fair. I bumped into a similar issue sometime ago and I ended up having to fix `any_one`.

mruberry replied:
That's true that we don't have good support for controlling the tolerances of tests whose dtype is determined by `any_one`, and you're right we have no mechanism for controlling it using decorators when the test has no dtype. We could add an 'any' or 'all' option to the toleranceOverride decorator to address this

srossross replied:
resolved in another PR on the stack!

lezcano commented on test/test_ops.py:
```diff
@@ -160,6 +161,34 @@ def test_numpy_ref(self, device, dtype, op):
+        if dtype not in op.dtypesIfCUDA:
+            raise unittest.SkipTest("This test requires both cuda and cpu support")
+
+        def to_cpu(arg):
+            if isinstance(arg, torch.dtype):
+                return arg
+            return arg.to(device='cpu')
```
This is a very intriguing function. Are the arguments passed to this function just `dtype` or `Tensor`? In any case, a more natural way of writing this I think it'd be
```suggestion
        def to_cpu(arg):
            if isinstance(arg, Tensor):
                return arg.to(device="cpu")
            return arg
```

srossross replied:
updated

lezcano reviewed (commented):
I just left a couple comments, but overall looks good.

Now, why didn't this test detect https://github.com/pytorch/pytorch/issues/85960? Is it just due to lack of sample coverage?

In any case, this is pretty much ready. Could you have a look @mruberry?

srossross committed:
Update on "Added new test test_compare_cpu that checks if cpu and gpu results are consistent"




[ghstack-poisoned]

srossross commented:
> Now, why didn't this test detect https://github.com/pytorch/pytorch/issues/85960? Is it just due to lack of sample coverage?

@lezcano I rebased from master and added the new Expected failures

IvanYashchuk commented:
> Now, why didn't this test detect https://github.com/pytorch/pytorch/issues/85960? Is it just due to lack of sample coverage?

Because there was no OpInfo for `torch.native_batch_norm` until today, it's not the same as `functional.batch_norm`.

amjames reviewed (dismissed):
See inline comments for a fix to the selection of dtype issue. Otherwise looks good.

mruberry commented on test/test_ops.py:
```diff
@@ -96,6 +96,7 @@
     )
 )
 _ops_and_refs = op_db + python_ref_db
+_ops_and_refs_with_no_numpy_ref = [op for op in _ops_and_refs if op.ref is None]
```
Add a comment explaining this list

srossross replied:
added

mruberry commented on test/test_ops.py:
```diff
@@ -165,6 +166,34 @@ def test_numpy_ref(self, device, dtype, op):
+    @ops(_ops_and_refs_with_no_numpy_ref, dtypes=OpDTypes.any_one)
+    def test_compare_cpu(self, device, dtype, op):
+
+        if dtype not in op.dtypes:
```
Add a comment for this check

srossross replied:
added

mruberry commented on test/test_ops.py:
```diff
@@ -165,6 +166,34 @@ def test_numpy_ref(self, device, dtype, op):
+            cuda_results = op(sample.input, *sample.args, **sample.kwargs)
+            cpu_results = op(cpu_sample.input, *cpu_sample.args, **cpu_sample.kwargs)
+
+            cuda_results = sample.output_process_fn_grad(cuda_results)
```
This is going to need a comment -- why are the results not compared directly?

lezcano replied:
`output_process_fn_grad` has a very unfortunate name. It should be `make_output_comparable` or smth lol
We use this function in `linalg` extensively to postprocess the inputs of functions that are not completely well-defined. Think svd and muliplying the singular vectors by `-1`. CPU and CUDA implementations of the SVD can return valid SVDs that are different. We use this function to compare them.

srossross replied:
comment added

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -10674,6 +10687,8 @@ def reference_flatten(input, start_dim=0, end_dim=-1):
                DecorateInfo(unittest.expectedFailure, 'TestGradients', 'test_forward_mode_AD'),
                # RuntimeError: deepEquals(input.iValue, deepCopiedInput) INTERNAL ASSERT FAILED
                DecorateInfo(unittest.expectedFailure, 'TestJit', 'test_variant_consistency_jit'),
+               # https://github.com/pytorch/pytorch/issues/85960
```
Nice issue citation

lezcano commented on test/test_ops.py:
```diff
@@ -165,6 +166,34 @@ def test_numpy_ref(self, device, dtype, op):
+                return arg.to(device='cpu')
+            return arg
+
+        samples = op.sample_inputs(device, dtype)
```
Actually, here we could use the `reference_inputs`. WDYT @mruberry?

mruberry replied:
Great catch, @lezcano. I agree that would be more consistent with similar tests

srossross replied:
added in https://github.com/pytorch/pytorch/pull/86462

srossross committed:
Update on "Added new test test_compare_cpu that checks if cpu and gpu results are consistent"




[ghstack-poisoned]

srossross committed:
Update on "Added new test test_compare_cpu that checks if cpu and gpu results are consistent"




[ghstack-poisoned]

srossross committed:
Update on "Added new test test_compare_cpu that checks if cpu and gpu results are consistent"




[ghstack-poisoned]

srossross committed:
Update on "Added new test test_compare_cpu that checks if cpu and gpu results are consistent"




[ghstack-poisoned]

mruberry commented on test/test_ops.py:
```diff
@@ -165,6 +170,34 @@ def test_numpy_ref(self, device, dtype, op):
+            cuda_results = op(sample.input, *sample.args, **sample.kwargs)
+            cpu_results = op(cpu_sample.input, *cpu_sample.args, **cpu_sample.kwargs)
+
+            # output_process_fn_grad has a very unfortunate name
```
Nice comment

mruberry commented:
The code looks good, @srossross. My final question is: how much time does this test take to run once it's switched to use reference inputs? Should we guard it with the `@slowtest` decorator? Finding the initial issues has been productive, but do you think this test is worth the cost of running all the time or only periodically?

lezcano commented:
FWIW, I vote to run this one just periodically, as it's not something that I expect it'll break very often

jjsjann123 commented:
> FWIW, I vote to run this one just periodically, as it's not something that I expect it'll break very often

What would be the CI label that would trigger this test? We do plan to fix cuda BN behavior and would like to verify the fix with it.

amjames commented:
If the test is only going to be run periodically, would it then be a better idea to include all dtypes from the intersection of CPU/GPU support?

mruberry commented:
> If the test is only going to be run periodically, would it then be a better idea to include all dtypes from the intersection of CPU/GPU support?

Maybe. But that seems like a lot even for a slow test. I wouldn't worry about that particular refinement.

srossross committed:
Update on "Added new test test_compare_cpu that checks if cpu and gpu results are consistent"




[ghstack-poisoned]

srossross committed:
Update on "Added new test test_compare_cpu that checks if cpu and gpu results are consistent"




[ghstack-poisoned]

srossross commented:
@mruberry I've done a few rebases to this stack - and each time - I've had to update the expectedFailures and or tolerances, I wonder if the `@slowTest` decorator could wait until after a period of stabilization? Please advise.

srossross committed:
Update on "Added new test test_compare_cpu that checks if cpu and gpu results are consistent"




[ghstack-poisoned]

srossross committed:
Update on "Added new test test_compare_cpu that checks if cpu and gpu results are consistent"




[ghstack-poisoned]

srossross committed:
Update on "Added new test test_compare_cpu that checks if cpu and gpu results are consistent"




[ghstack-poisoned]

lezcano commented on torch/testing/_internal/opinfo/definitions/fft.py:
```diff
@@ -217,7 +221,18 @@ def mt(shape, **kwargs):
+            ),
+            DecorateInfo(
+                toleranceOverride(
+                    {
+                        torch.float16: tol(atol=1e-4, rtol=1e-4),
+                        torch.complex64: tol(atol=1e-4, rtol=1e-4),
```
With the new tolerances, all these can be removed.

srossross replied:
done

lezcano reviewed (approved):
This PR will need a clean-up, but otherwise LGTM!

srossross committed:
Update on "Added new test test_compare_cpu that checks if cpu and gpu results are consistent"




[ghstack-poisoned]

srossross committed:
Update on "Added new test test_compare_cpu that checks if cpu and gpu results are consistent"




[ghstack-poisoned]

mruberry reviewed (approved):
Cool! LGTM!

srossross commented:
@pytorchbot merge
